data.go

```diff
@@ -85,86 +64,287 @@ func setDB(db apid.DB) {
+		for _, columnName := range orderedColumns {
+			//use Value so that stmt exec does not complain about common.ColumnVal being a struct
+			values = append(values, row[columnName].Value)
+		}
+	}
+
+	//create prepared statement from existing template statement
+	_, err = prep.Exec(values...)
 
 	if err != nil {
-		log.Errorf("insert APID_CLUSTER failed: %v", err)
+		log.Errorf("INSERT Fail [%s] values=%v error=[%v]", sql, values, err)
+		return false
+	} else {
+		log.Debugf("INSERT Success [%s] values=%v", sql, values)
 	}
 
-	return err
+	return true
+}
+
+func getValueListFromKeys(row common.Row, pkeys []string) []interface{} {
+	var values = make([]interface{}, len(pkeys))
+	for i, pkey := range pkeys {
+		if row[pkey] == nil {
+			values[i] = nil
+		} else {
+			values[i] = row[pkey].Value
+		}
+	}
+	return values
 }
 
-func insertDataScope(ds dataDataScope, txn *sql.Tx) error {
+func _delete(tableName string, rows []common.Row, txn *sql.Tx) bool {
+	pkeys, err := getPkeysForTable(tableName)
+	sort.Strings(pkeys)
+	if len(pkeys) == 0 || err != nil {
+		log.Errorf("DELETE No primary keys found for table. %s", tableName)
+		return false
+	} else if len(rows) == 0 {
+		log.Errorf("No rows found for table.", tableName)
+		return false
+	} else {
+		sql := buildDeleteSql(tableName, rows[0], pkeys)
+		prep, err := txn.Prepare(sql)
+		if err != nil {
+			log.Errorf("DELETE Fail to prep statement [%s] error=[%v]", sql, err)
+			return false
+		}
+		defer prep.Close()
+		for _, row := range rows {
+			values := getValueListFromKeys(row, pkeys)
+			// delete prepared statement from existing template statement
+			res, err := txn.Stmt(prep).Exec(values...)
+			if err != nil {
+				log.Errorf("DELETE Fail [%s] values=%v error=[%v]", sql, values, err)
+				return false
+			} else {
+				affected, err := res.RowsAffected()
+				if err == nil && affected != 0 {
+					log.Debugf("DELETE Success [%s] values=%v", sql, values)
+				} else if err == nil && affected == 0 {
+					log.Errorf("Entry not found [%s] values=%v. Nothing to delete.", sql, values)
+					return false
+				} else {
+					log.Errorf("DELETE Failed [%s] values=%v error=[%v]", sql, values, err)
+					return false
+				}
+			}
+		}
+		return true
+	}
+}
 
-	log.Debugf("insert DATA_SCOPE: %v", ds)
+// Syntax "DELETE FROM Obj WHERE key1=$1 AND key2=$2 ... ;"
+func buildDeleteSql(tableName string, row common.Row, pkeys []string) string {
 
-	//replace to accomodate same snapshot txid
-	stmt, err := txn.Prepare(`
-	REPLACE INTO DATA_SCOPE
-		(id, apid_cluster_id, scope, org,
-		env, created, created_by, updated,
-		updated_by)
-	VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9);
-	`)
-	if err != nil {
-		log.Errorf("insert DATA_SCOPE failed: %v", err)
-		return err
+	var wherePlaceholders []string
+	i := 1
+	if row == nil {
+		return ""
+	}
+	normalizedTableName := normalizeTableName(tableName)
+
+	for _, pk := range pkeys {
+		wherePlaceholders = append(wherePlaceholders, fmt.Sprintf("%s=$%v", pk, i))
```

Comment:
Why not:
for i, pk := range pkeys {
    wherePlaceholders = append(wherePlaceholders, fmt.Sprintf("%s=$%v", pk, i+1))
}